src/modules/secrets/secrets.service.ts

```diff
@@ -0,0 +1,125 @@
+        throw new InternalServerErrorException('Failed to create secret.');
+      }
+      return createdSecret;
+    } catch {
+      throw new BadRequestException('Could not create secret');
+    }
+  }
+
+  async findAllSecretsByAccount(accountId: number) {
+    try {
+      const secrets = await this.secretsDAL.findAllSecrets(accountId);
+      if (secrets.length === 0) {
+        throw new NotFoundException('No secrets found for this account.');
+      }
+      const account = await this.accountsDAL.findAccountById(accountId);
+    if (!account) {
+      throw new NotFoundException('Account not found.');
+    }
+      const decryptedSecrets = secrets.map(secret => {
+        const decryptedValue = decrypt(secret.value, account.password);
+        return { ...secret, value: decryptedValue };
+      }); 
+      return decryptedSecrets;
+    } catch {
+      throw new InternalServerErrorException('Failed to retrieve secrets.');
+    }
+  }
+
+  async findSecretByIdAndAccount(accountId: number, secretId: number) {
+    try {
+      const secret = await this.secretsDAL.findSecretById(secretId, accountId);
+      if (!secret) {
+        throw new NotFoundException(
+          'Secret not found or does not belong to this account.',
+        );
+      }
+      const account = await this.accountsDAL.findAccountById(accountId);
+    if (!account) {
+      throw new NotFoundException('Account not found.');
+    }
+        const decryptedValue = decrypt(secret.value, account.password);
+        return { ...secret, value: decryptedValue };
+    } catch {
+      throw new InternalServerErrorException('Failed to retrieve secret.');
+    }
+  }
+
+  async updateSecret(
+    accountId: number,
+    secretId: number,
+    updateSecretDto: UpdateSecretsDto,
+  ) {
+    try {
+      const secret = await this.findSecretByIdAndAccount(accountId, secretId);
+      if (!secret) {
+        throw new NotFoundException('Secret not found for this account.');
+      }
+      const account = await this.accountsDAL.findAccountById(accountId);
+      if (!account) {
+        throw new NotFoundException('Account not found.');
+      }
+      if (updateSecretDto.value) {
+        updateSecretDto.value = encrypt(updateSecretDto.value, account.password);
+      }
+      return await this.secretsDAL.updateSecret(
+        secretId,
+        updateSecretDto,
+        accountId,
+      );
+    } catch (error) {
+      if (error instanceof NotFoundException) {
+        throw error;
+      }
+      throw new InternalServerErrorException('Failed to update secret.');
+    }
+  }
+
+  async deleteSecret(accountId: number, secretId: number) {
+    try {
+      const secret = await this.findSecretByIdAndAccount(accountId, secretId);
+      if (!secret) {
+        throw new NotFoundException('Secret not found for this account.');
+      }
+
+      return await this.secretsDAL.deleteSecret(secretId, accountId);
+    } catch (error) {
+      if (error instanceof NotFoundException) {
```

Comment:
@danielaKokoneshi309 If prisma handles the not-found scenario for us, this could be simplified to the following:

```
try {
  // Logic here
} catch (CustomNotFoundErrorFromPrisma e) {
  throw new NotFoundException(); // You convert a DAL exception to a domain-layer exception. It does not change much, but it's all about layers (we don't want to handle a DAL error inside the controller, since we have services in between)
}
```